ICE: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs: unexpected path

<!--
Thank you for finding an Internal Compiler Error! 🧊  If possible, try to provide
a minimal verifiable example. You can read "Rust Bug Minimization Patterns" for
how to create smaller examples.

http://blog.pnkfx.org/blog/2019/11/18/rust-bug-minimization-patterns/

-->

### Code
kinda garbish-y but it reproduces the ICE :shrug: 
```Rust
use std::path::Path;

pub use self::imp::Lock;

#[cfg(unix)]
mod imp {
    use std::ffi::{CString, OsStr};
    use std::os::unix::prelude::*;
    use std::path::Path;
    use std::io;
    use libc;

  

    #[derive(Debug)]
    pub struct Lock {
        fd: libc::c_int,
    }

    impl Lock {
        pub fn new(p: &Path,
                   wait: bool,
                   create: bool,
                   exclusive: bool)
                   -> io::Result<Lock> {
            let os: &OsStr = p.as_ref();
            let buf = CString::new(os.as_bytes()).unwrap();
            let open_flags = if create {
                libc::O_RDWR | libc::O_CREAT
            } else {
                libc::O_RDWR
            };

            let fd = unsafe {
                libc::open(buf.as_ptr(), open_flags,
                           libc::S_IRWXU as libc::c_int)
            };

            let cmd = if wait { os::F_SETLKW } else { os::F_SETLK };
            let ret = unsafe {
                
            };
            if ret == -1 {
                let err = io::Error::last_os_error();
                unsafe { libc::close(fd); }
                Err(err)
            } else {
                Ok(Lock { fd: fd })
            }

        }
    }

}
```


### Meta
<!--
If you're using the stable version of the compiler, you should also check if the
bug also exists in the beta or nightly versions.
-->

`rustc --version --verbose`:

IMPORTANT: this reproduces with MASTER but not yet with NIGHTLY `rustc 1.63.0-nightly (e71440575 2022-06-02)`
```
rustc 1.63.0-nightly (e40d5e83d 2022-06-03)
binary: rustc
commit-hash: e40d5e83dc133d093c22c7ff016b10daa4f40dcf
commit-date: 2022-06-03
host: x86_64-unknown-linux-gnu
release: 1.63.0-nightly
LLVM version: 14.0.4
```

### Error output

```
error[E0432]: unresolved import `libc`
  --> ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:11:9
   |
11 |     use libc;
   |         ^^^^ no `libc` in the root

error[E0433]: failed to resolve: use of undeclared crate or module `os`
  --> ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:39:33
   |
39 |             let cmd = if wait { os::F_SETLKW } else { os::F_SETLK };
   |                                 ^^ use of undeclared crate or module `os`

error[E0433]: failed to resolve: use of undeclared crate or module `os`
  --> ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:39:55
   |
39 |             let cmd = if wait { os::F_SETLKW } else { os::F_SETLK };
   |                                                       ^^ use of undeclared crate or module `os`

warning: unused import: `std::path::Path`
 --> ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:1:5
  |
1 | use std::path::Path;
  |     ^^^^^^^^^^^^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

error[E0601]: `main` function not found in crate `510c9ceef0960a8c8b6a20e3625ee6c2e10b4052`
  --> ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:54:2
   |
54 | }
   |  ^ consider adding a `main` function to `./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs`

error[E0308]: mismatched types
  --> ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:43:23
   |
43 |             if ret == -1 {
   |                       ^^ expected `()`, found integer

error: internal compiler error: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:856:33: unexpected path: def=alloc::ffi::c_str::CString substs=[] path=Path { span: ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:27:23: 27:30 (#0), res: Def(TyAlias, DefId(1:2774 ~ std[4f4f]::ffi::CString)), segments: [PathSegment { ident: CString#0, hir_id: Some(HirId { owner: DefId(0:28 ~ 510c9ceef0960a8c8b6a20e3625ee6c2e10b4052[d1fc]::imp::{impl#0}::new), local_id: 22 }), res: Some(Err), args: None, infer_args: true }] }

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/compiler/rustc_errors/src/lib.rs:1335:9
```

<!--
Include a backtrace in the code block by setting `RUST_BACKTRACE=1` in your
environment. E.g. `RUST_BACKTRACE=1 cargo build`.
-->
<details><summary><strong>Backtrace</strong></summary>
<p>

```
error: internal compiler error: compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs:856:33: unexpected path: def=alloc::ffi::c_str::CString substs=[] path=Path { span: ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:27:23: 27:30 (#0), res: Def(TyAlias, DefId(1:2774 ~ std[4f4f]::ffi::CString)), segments: [PathSegment { ident: CString#0, hir_id: Some(HirId { owner: DefId(0:28 ~ 510c9ceef0960a8c8b6a20e3625ee6c2e10b4052[d1fc]::imp::{impl#0}::new), local_id: 22 }), res: Some(Err), args: None, infer_args: true }] }

thread 'rustc' panicked at 'Box<dyn Any>', /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/compiler/rustc_errors/src/lib.rs:1335:9
stack backtrace:
   0:     0x7f6b6ec9decd - std::backtrace_rs::backtrace::libunwind::trace::hfcd45b45ae488cd5
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f6b6ec9decd - std::backtrace_rs::backtrace::trace_unsynchronized::h405dac851f6cc6a8
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f6b6ec9decd - std::sys_common::backtrace::_print_fmt::h5f6c74a63449dd3f
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/sys_common/backtrace.rs:66:5
   3:     0x7f6b6ec9decd - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h7d487462c277bf24
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/sys_common/backtrace.rs:45:22
   4:     0x7f6b6ecf9bac - core::fmt::write::h832d0c8909aada18
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/core/src/fmt/mod.rs:1196:17
   5:     0x7f6b6ec8f5c1 - std::io::Write::write_fmt::h65c8c1822edc4894
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/io/mod.rs:1654:15
   6:     0x7f6b6eca0bb5 - std::sys_common::backtrace::_print::h560a0f4f262236c1
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/sys_common/backtrace.rs:48:5
   7:     0x7f6b6eca0bb5 - std::sys_common::backtrace::print::h0080326ddadae76c
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/sys_common/backtrace.rs:35:9
   8:     0x7f6b6eca0bb5 - std::panicking::default_hook::{{closure}}::hc726d21ea0224885
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/panicking.rs:295:22
   9:     0x7f6b6eca08d6 - std::panicking::default_hook::hdcc52b852804bb0b
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/panicking.rs:314:9
  10:     0x7f6b6f4f9f11 - rustc_driver[c7c8f2fedff45b54]::DEFAULT_HOOK::{closure#0}::{closure#0}
  11:     0x7f6b6eca128a - std::panicking::rust_panic_with_hook::h48930cf2efbb5ae2
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/panicking.rs:702:17
  12:     0x7f6b705e7ae1 - std[4f4f1627d3880017]::panicking::begin_panic::<rustc_errors[65331a6f7e432222]::ExplicitBug>::{closure#0}
  13:     0x7f6b705e7556 - std[4f4f1627d3880017]::sys_common::backtrace::__rust_end_short_backtrace::<std[4f4f1627d3880017]::panicking::begin_panic<rustc_errors[65331a6f7e432222]::ExplicitBug>::{closure#0}, !>
  14:     0x7f6b7054ecf6 - std[4f4f1627d3880017]::panicking::begin_panic::<rustc_errors[65331a6f7e432222]::ExplicitBug>
  15:     0x7f6b70572b26 - std[4f4f1627d3880017]::panic::panic_any::<rustc_errors[65331a6f7e432222]::ExplicitBug>
  16:     0x7f6b705722e5 - <rustc_errors[65331a6f7e432222]::HandlerInner>::bug::<&alloc[31c74b9f3315cbc6]::string::String>
  17:     0x7f6b70571dc0 - <rustc_errors[65331a6f7e432222]::Handler>::bug::<&alloc[31c74b9f3315cbc6]::string::String>
  18:     0x7f6b705e4acd - rustc_middle[83233ac63f698a09]::ty::context::tls::with_opt::<rustc_middle[83233ac63f698a09]::util::bug::opt_span_bug_fmt<rustc_span[b73b5b7a1d5e4b08]::span_encoding::Span>::{closure#0}, ()>
  19:     0x7f6b705e4bc6 - rustc_middle[83233ac63f698a09]::util::bug::opt_span_bug_fmt::<rustc_span[b73b5b7a1d5e4b08]::span_encoding::Span>
  20:     0x7f6b705e4b43 - rustc_middle[83233ac63f698a09]::util::bug::bug_fmt
  21:     0x7f6b704a82ba - <rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor>::path_inferred_subst_iter
  22:     0x7f6b704a8833 - <rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[a9ca4d4e6c54b339]::intravisit::Visitor>::visit_expr
  23:     0x7f6b704a86cc - <rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[a9ca4d4e6c54b339]::intravisit::Visitor>::visit_expr
  24:     0x7f6b704beb92 - rustc_hir[a9ca4d4e6c54b339]::intravisit::walk_expr::<rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor>
  25:     0x7f6b704a86de - <rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[a9ca4d4e6c54b339]::intravisit::Visitor>::visit_expr
  26:     0x7f6b704bb80c - rustc_hir[a9ca4d4e6c54b339]::intravisit::walk_local::<rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor>
  27:     0x7f6b704a8348 - <rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[a9ca4d4e6c54b339]::intravisit::Visitor>::visit_local
  28:     0x7f6b704bb657 - rustc_hir[a9ca4d4e6c54b339]::intravisit::walk_block::<rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor>
  29:     0x7f6b704a86de - <rustc_infer[78bfaa2976d36a75]::infer::error_reporting::need_type_info::FindInferSourceVisitor as rustc_hir[a9ca4d4e6c54b339]::intravisit::Visitor>::visit_expr
  30:     0x7f6b70460f84 - <rustc_infer[78bfaa2976d36a75]::infer::InferCtxt>::emit_inference_failure_err
  31:     0x7f6b703b30c2 - <rustc_infer[78bfaa2976d36a75]::infer::InferCtxt as rustc_trait_selection[ae7a9636b275321f]::traits::error_reporting::InferCtxtPrivExt>::maybe_report_ambiguity
  32:     0x7f6b703a6c87 - <rustc_infer[78bfaa2976d36a75]::infer::InferCtxt as rustc_trait_selection[ae7a9636b275321f]::traits::error_reporting::InferCtxtExt>::report_fulfillment_errors
  33:     0x7f6b70af7af3 - <rustc_infer[78bfaa2976d36a75]::infer::InferCtxtBuilder>::enter::<&rustc_middle[83233ac63f698a09]::ty::context::TypeckResults, <rustc_typeck[412d93c49a5cfb7d]::check::inherited::InheritedBuilder>::enter<rustc_typeck[412d93c49a5cfb7d]::check::typeck_with_fallback<rustc_typeck[412d93c49a5cfb7d]::check::typeck::{closure#0}>::{closure#1}, &rustc_middle[83233ac63f698a09]::ty::context::TypeckResults>::{closure#0}>
  34:     0x7f6b70aa6efa - rustc_typeck[412d93c49a5cfb7d]::check::typeck
  35:     0x7f6b7109bbb4 - rustc_query_system[cd54684135f7aede]::query::plumbing::try_execute_query::<rustc_query_impl[5d43592eeec5f58a]::plumbing::QueryCtxt, rustc_query_system[cd54684135f7aede]::query::caches::DefaultCache<rustc_span[b73b5b7a1d5e4b08]::def_id::LocalDefId, &rustc_middle[83233ac63f698a09]::ty::context::TypeckResults>>
  36:     0x7f6b7107623e - <rustc_query_impl[5d43592eeec5f58a]::Queries as rustc_middle[83233ac63f698a09]::ty::query::QueryEngine>::typeck
  37:     0x7f6b70b55ee8 - <rustc_middle[83233ac63f698a09]::hir::map::Map>::par_body_owners::<rustc_typeck[412d93c49a5cfb7d]::check::typeck_item_bodies::{closure#0}>
  38:     0x7f6b71936a7c - rustc_typeck[412d93c49a5cfb7d]::check::typeck_item_bodies
  39:     0x7f6b71c303ac - rustc_query_system[cd54684135f7aede]::query::plumbing::try_execute_query::<rustc_query_impl[5d43592eeec5f58a]::plumbing::QueryCtxt, rustc_query_system[cd54684135f7aede]::query::caches::DefaultCache<(), ()>>
  40:     0x7f6b71c5a8b1 - rustc_query_system[cd54684135f7aede]::query::plumbing::get_query::<rustc_query_impl[5d43592eeec5f58a]::queries::typeck_item_bodies, rustc_query_impl[5d43592eeec5f58a]::plumbing::QueryCtxt>
  41:     0x7f6b7196d3f3 - <rustc_session[3dcad23083b71834]::session::Session>::time::<(), rustc_typeck[412d93c49a5cfb7d]::check_crate::{closure#7}>
  42:     0x7f6b7194a77b - rustc_typeck[412d93c49a5cfb7d]::check_crate
  43:     0x7f6b716e3a57 - rustc_interface[d21641337a9fbe3]::passes::analysis
  44:     0x7f6b71c2649f - rustc_query_system[cd54684135f7aede]::query::plumbing::try_execute_query::<rustc_query_impl[5d43592eeec5f58a]::plumbing::QueryCtxt, rustc_query_system[cd54684135f7aede]::query::caches::DefaultCache<(), core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>>>
  45:     0x7f6b71c6ce2e - rustc_query_system[cd54684135f7aede]::query::plumbing::get_query::<rustc_query_impl[5d43592eeec5f58a]::queries::analysis, rustc_query_impl[5d43592eeec5f58a]::plumbing::QueryCtxt>
  46:     0x7f6b716c6a07 - <rustc_interface[d21641337a9fbe3]::passes::QueryContext>::enter::<rustc_driver[c7c8f2fedff45b54]::run_compiler::{closure#1}::{closure#2}::{closure#3}, core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>>
  47:     0x7f6b716b0cc8 - <rustc_interface[d21641337a9fbe3]::interface::Compiler>::enter::<rustc_driver[c7c8f2fedff45b54]::run_compiler::{closure#1}::{closure#2}, core[e20aa153a98323b0]::result::Result<core[e20aa153a98323b0]::option::Option<rustc_interface[d21641337a9fbe3]::queries::Linker>, rustc_errors[65331a6f7e432222]::ErrorGuaranteed>>
  48:     0x7f6b716d44df - rustc_span[b73b5b7a1d5e4b08]::with_source_map::<core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>, rustc_interface[d21641337a9fbe3]::interface::create_compiler_and_run<core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>, rustc_driver[c7c8f2fedff45b54]::run_compiler::{closure#1}>::{closure#1}>
  49:     0x7f6b716b18f2 - <scoped_tls[f8b3eb90e665007e]::ScopedKey<rustc_span[b73b5b7a1d5e4b08]::SessionGlobals>>::set::<rustc_interface[d21641337a9fbe3]::interface::run_compiler<core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>, rustc_driver[c7c8f2fedff45b54]::run_compiler::{closure#1}>::{closure#0}, core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>>
  50:     0x7f6b716c708f - std[4f4f1627d3880017]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[d21641337a9fbe3]::util::run_in_thread_pool_with_globals<rustc_interface[d21641337a9fbe3]::interface::run_compiler<core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>, rustc_driver[c7c8f2fedff45b54]::run_compiler::{closure#1}>::{closure#0}, core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>>::{closure#0}, core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>>
  51:     0x7f6b716c71c9 - <<std[4f4f1627d3880017]::thread::Builder>::spawn_unchecked_<rustc_interface[d21641337a9fbe3]::util::run_in_thread_pool_with_globals<rustc_interface[d21641337a9fbe3]::interface::run_compiler<core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>, rustc_driver[c7c8f2fedff45b54]::run_compiler::{closure#1}>::{closure#0}, core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>>::{closure#0}, core[e20aa153a98323b0]::result::Result<(), rustc_errors[65331a6f7e432222]::ErrorGuaranteed>>::{closure#1} as core[e20aa153a98323b0]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  52:     0x7f6b6ecab1b3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h3752e20e7af2922b
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/alloc/src/boxed.rs:1951:9
  53:     0x7f6b6ecab1b3 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::hea44272e5391750e
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/alloc/src/boxed.rs:1951:9
  54:     0x7f6b6ecab1b3 - std::sys::unix::thread::Thread::new::thread_start::h6d31b3f92a2b25ea
                               at /rustc/e40d5e83dc133d093c22c7ff016b10daa4f40dcf/library/std/src/sys/unix/thread.rs:108:17
  55:     0x7f6b6ea835c2 - start_thread
  56:     0x7f6b6eb08584 - __clone
  57:                0x0 - <unknown>

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.63.0-nightly (e40d5e83d 2022-06-03) running on x86_64-unknown-linux-gnu

query stack during panic:
#0 [typeck] type-checking `imp::<impl at ./510c9ceef0960a8c8b6a20e3625ee6c2e10b4052.rs:20:5: 52:6>::new`
#1 [typeck_item_bodies] type-checking all item bodies
#2 [analysis] running analysis passes on this crate
end of query stack
error: aborting due to 6 previous errors; 1 warning emitted

Some errors have detailed explanations: E0308, E0432, E0433, E0601.
For more information about an error, try `rustc --explain E0308`.
```

</p>
</details>

--- a/compiler/rustc_infer/src/infer/error_reporting/mod.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/mod.rs
@@ -67,10 +67,8 @@ use rustc_hir::{Item, ItemKind, Node};
 use rustc_middle::dep_graph::DepContext;
 use rustc_middle::ty::print::with_no_trimmed_paths;
 use rustc_middle::ty::{
-    self,
-    error::TypeError,
-    subst::{GenericArgKind, Subst, SubstsRef},
-    Binder, EarlyBinder, List, Region, Ty, TyCtxt, TypeFoldable, TypeSuperFoldable,
+    self, error::TypeError, Binder, List, Region, Subst, Ty, TyCtxt, TypeFoldable,
+    TypeSuperFoldable,
 };
 use rustc_span::{sym, symbol::kw, BytePos, DesugaringKind, Pos, Span};
 use rustc_target::spec::abi;
@@ -926,10 +924,13 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
         mut t1_out: &mut DiagnosticStyledString,
         mut t2_out: &mut DiagnosticStyledString,
         path: String,
-        sub: ty::subst::SubstsRef<'tcx>,
+        sub: &'tcx [ty::GenericArg<'tcx>],
         other_path: String,
         other_ty: Ty<'tcx>,
     ) -> Option<()> {
+        // FIXME/HACK: Go back to `SubstsRef` to use its inherent methods,
+        // ideally that shouldn't be necessary.
+        let sub = self.tcx.intern_substs(sub);
         for (i, ta) in sub.types().enumerate() {
             if ta == other_ty {
                 self.highlight_outer(&mut t1_out, &mut t2_out, path, sub, i, other_ty);
@@ -958,45 +959,6 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             value.push_normal(", ");
             other_value.push_normal(", ");
         }
-    }
-
-    /// For generic types with parameters with defaults, remove the parameters corresponding to
-    /// the defaults. This repeats a lot of the logic found in `ty::print::pretty`.
-    fn strip_generic_default_params(
-        &self,
-        def_id: DefId,
-        substs: ty::subst::SubstsRef<'tcx>,
-    ) -> SubstsRef<'tcx> {
-        let generics = self.tcx.generics_of(def_id);
-        let mut num_supplied_defaults = 0;
-
-        let default_params = generics.params.iter().rev().filter_map(|param| match param.kind {
-            ty::GenericParamDefKind::Type { has_default: true, .. } => Some(param.def_id),
-            ty::GenericParamDefKind::Const { has_default: true } => Some(param.def_id),
-            _ => None,
-        });
-        for (def_id, actual) in iter::zip(default_params, substs.iter().rev()) {
-            match actual.unpack() {
-                GenericArgKind::Const(c) => {
-                    if EarlyBinder(self.tcx.const_param_default(def_id)).subst(self.tcx, substs)
-                        != c
-                    {
-                        break;
-                    }
-                }
-                GenericArgKind::Type(ty) => {
-                    if self.tcx.bound_type_of(def_id).subst(self.tcx, substs) != ty {
-                        break;
-                    }
-                }
-                _ => break,
-            }
-            num_supplied_defaults += 1;
-        }
-        let len = generics.params.len();
-        let mut generics = generics.clone();
-        generics.params.truncate(len - num_supplied_defaults);
-        substs.truncate_to(self.tcx, &generics)
     }
 
     /// Given two `fn` signatures highlight only sub-parts that are different.
@@ -1156,8 +1118,10 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
             (&ty::Adt(def1, sub1), &ty::Adt(def2, sub2)) => {
                 let did1 = def1.did();
                 let did2 = def2.did();
-                let sub_no_defaults_1 = self.strip_generic_default_params(did1, sub1);
-                let sub_no_defaults_2 = self.strip_generic_default_params(did2, sub2);
+                let sub_no_defaults_1 =
+                    self.tcx.generics_of(did1).own_substs_no_defaults(self.tcx, sub1);
+                let sub_no_defaults_2 =
+                    self.tcx.generics_of(did2).own_substs_no_defaults(self.tcx, sub2);
                 let mut values = (DiagnosticStyledString::new(), DiagnosticStyledString::new());
                 let path1 = self.tcx.def_path_str(did1);
                 let path2 = self.tcx.def_path_str(did2);

--- a/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
@@ -11,7 +11,7 @@ use rustc_middle::infer::unify_key::ConstVariableOriginKind;
 use rustc_middle::ty::adjustment::{Adjust, Adjustment, AutoBorrow, AutoBorrowMutability};
 use rustc_middle::ty::print::{FmtPrinter, PrettyPrinter, Print, Printer};
 use rustc_middle::ty::subst::{GenericArg, GenericArgKind, Subst, SubstsRef};
-use rustc_middle::ty::{self, DefIdTree, GenericParamDefKind, InferConst};
+use rustc_middle::ty::{self, DefIdTree, InferConst};
 use rustc_middle::ty::{Ty, TyCtxt, TypeckResults};
 use rustc_span::symbol::{kw, Ident};
 use rustc_span::{BytePos, Span};
@@ -958,26 +958,8 @@ impl<'a, 'tcx> Visitor<'tcx> for FindInferSourceVisitor<'a, 'tcx> {
                 generics.own_substs(substs).iter().position(|&arg| self.generic_arg_is_target(arg))
             {
                 let substs = self.infcx.resolve_vars_if_possible(substs);
-                let num_args = generics
-                    .params
-                    .iter()
-                    .rev()
-                    .filter(|&p| !matches!(p.kind, GenericParamDefKind::Lifetime))
-                    .skip_while(|&param| {
-                        if let Some(default) = param.default_value(tcx) {
-                            // FIXME: Using structural comparisions has a bunch of false negatives.
-                            //
-                            // We should instead try to replace inference variables with placeholders and
-                            // then use `infcx.can_eq`. That probably should be a separate method
-                            // generally used during error reporting.
-                            default.subst(tcx, substs) == substs[param.index as usize]
-                        } else {
-                            false
-                        }
-                    })
-                    .count();
-                let generic_args =
-                    &generics.own_substs(substs)[generics.own_counts().lifetimes..][..num_args];
+                let generic_args = &generics.own_substs_no_defaults(tcx, substs)
+                    [generics.own_counts().lifetimes..];
                 let span = match expr.kind {
                     ExprKind::MethodCall(path, _, _) => path.ident.span,
                     _ => expr.span,

--- a/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
@@ -2,6 +2,7 @@ use crate::infer::type_variable::TypeVariableOriginKind;
 use crate::infer::InferCtxt;
 use rustc_errors::{pluralize, struct_span_err, Applicability, DiagnosticBuilder, ErrorGuaranteed};
 use rustc_hir as hir;
+use rustc_hir::def::Res;
 use rustc_hir::def::{CtorOf, DefKind, Namespace};
 use rustc_hir::def_id::DefId;
 use rustc_hir::intravisit::{self, Visitor};
@@ -853,12 +854,20 @@ impl<'a, 'tcx> FindInferSourceVisitor<'a, 'tcx> {
                             hir::TyKind::Path(hir::QPath::Resolved(_self_ty, path)),
                         ) => {
                             if tcx.res_generics_def_id(path.res) != Some(def.did()) {
-                                bug!(
-                                    "unexpected path: def={:?} substs={:?} path={:?}",
-                                    def,
-                                    substs,
-                                    path,
-                                );
+                                match path.res {
+                                    Res::Def(DefKind::TyAlias, _) => {
+                                        // FIXME: Ideally we should support this. For that
+                                        // we have to map back from the self type to the
+                                        // type alias though. That's difficult.
+                                        //
+                                        // See the `need_type_info/type-alias.rs` test for
+                                        // some examples.
+                                    }
+                                    _ => warn!(
+                                        "unexpected path: def={:?} substs={:?} path={:?}",
+                                        def, substs, path,
+                                    ),
+                                }
                             } else {
                                 return Box::new(
                                     self.resolved_path_inferred_subst_iter(path, substs)

--- a/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
+++ b/compiler/rustc_infer/src/infer/error_reporting/need_type_info.rs
@@ -863,6 +863,9 @@ impl<'a, 'tcx> FindInferSourceVisitor<'a, 'tcx> {
                                         // See the `need_type_info/type-alias.rs` test for
                                         // some examples.
                                     }
+                                    // There cannot be inference variables in the self type,
+                                    // so there's nothing for us to do here.
+                                    Res::SelfTy { .. } => {}
                                     _ => warn!(
                                         "unexpected path: def={:?} substs={:?} path={:?}",
                                         def, substs, path,

--- a/compiler/rustc_middle/src/ty/generics.rs
+++ b/compiler/rustc_middle/src/ty/generics.rs
@@ -230,6 +230,8 @@ impl<'tcx> Generics {
 
     /// Returns the substs corresponding to the generic parameters
     /// of this item, excluding `Self`.
+    ///
+    /// **This should only be used for diagnostics purposes.**
     pub fn own_substs_no_defaults(
         &'tcx self,
         tcx: TyCtxt<'tcx>,
@@ -261,6 +263,8 @@ impl<'tcx> Generics {
     }
 
     /// Returns the substs corresponding to the generic parameters of this item, excluding `Self`.
+    ///
+    /// **This should only be used for diagnostics purposes.**
     pub fn own_substs(
         &'tcx self,
         substs: &'tcx [ty::GenericArg<'tcx>],

--- a/compiler/rustc_middle/src/ty/generics.rs
+++ b/compiler/rustc_middle/src/ty/generics.rs
@@ -228,8 +228,43 @@ impl<'tcx> Generics {
         })
     }
 
+    /// Returns the substs corresponding to the generic parameters
+    /// of this item, excluding `Self`.
+    pub fn own_substs_no_defaults(
+        &'tcx self,
+        tcx: TyCtxt<'tcx>,
+        substs: &'tcx [ty::GenericArg<'tcx>],
+    ) -> &'tcx [ty::GenericArg<'tcx>] {
+        let mut own_params = self.parent_count..self.count();
+        if self.has_self && self.parent.is_none() {
+            own_params.start = 1;
+        }
+
+        // Filter the default arguments.
+        //
+        // This currently uses structural equality instead
+        // of semantic equivalance. While not ideal, that's
+        // good enough for now as this should only be used
+        // for diagnostics anyways.
+        own_params.end -= self
+            .params
+            .iter()
+            .rev()
+            .take_while(|param| {
+                param.default_value(tcx).map_or(false, |default| {
+                    default.subst(tcx, substs) == substs[param.index as usize]
+                })
+            })
+            .count();
+
+        &substs[own_params]
+    }
+
     /// Returns the substs corresponding to the generic parameters of this item, excluding `Self`.
-    pub fn own_substs(&'tcx self, substs: SubstsRef<'tcx>) -> &'tcx [ty::GenericArg<'tcx>] {
+    pub fn own_substs(
+        &'tcx self,
+        substs: &'tcx [ty::GenericArg<'tcx>],
+    ) -> &'tcx [ty::GenericArg<'tcx>] {
         let own = &substs[self.parent_count..][..self.params.len()];
         if self.has_self && self.parent.is_none() { &own[1..] } else { &own }
     }

--- a/compiler/rustc_middle/src/ty/mod.rs
+++ b/compiler/rustc_middle/src/ty/mod.rs
@@ -22,7 +22,6 @@ use crate::mir::{Body, GeneratorLayout};
 use crate::traits::{self, Reveal};
 use crate::ty;
 use crate::ty::fast_reject::SimplifiedType;
-use crate::ty::subst::{GenericArg, InternalSubsts, Subst, SubstsRef};
 use crate::ty::util::Discr;
 pub use adt::*;
 pub use assoc::*;
@@ -44,6 +43,7 @@ use rustc_session::cstore::CrateStoreDyn;
 use rustc_span::symbol::{kw, sym, Ident, Symbol};
 use rustc_span::Span;
 use rustc_target::abi::Align;
+pub use subst::*;
 pub use vtable::*;
 
 use std::fmt::Debug;

--- a/compiler/rustc_middle/src/ty/print/mod.rs
+++ b/compiler/rustc_middle/src/ty/print/mod.rs
@@ -149,7 +149,7 @@ pub trait Printer<'tcx>: Sized {
                         // on top of the same path, but without its own generics.
                         _ => {
                             if !generics.params.is_empty() && substs.len() >= generics.count() {
-                                let args = self.generic_args_to_print(generics, substs);
+                                let args = generics.own_substs_no_defaults(self.tcx(), substs);
                                 return self.path_generic_args(
                                     |cx| cx.print_def_path(def_id, parent_substs),
                                     args,
@@ -182,43 +182,6 @@ pub trait Printer<'tcx>: Sized {
                 )
             }
         }
-    }
-
-    fn generic_args_to_print(
-        &self,
-        generics: &'tcx ty::Generics,
-        substs: &'tcx [GenericArg<'tcx>],
-    ) -> &'tcx [GenericArg<'tcx>] {
-        let mut own_params = generics.parent_count..generics.count();
-
-        // Don't print args for `Self` parameters (of traits).
-        if generics.has_self && own_params.start == 0 {
-            own_params.start = 1;
-        }
-
-        // Don't print args that are the defaults of their respective parameters.
-        own_params.end -= generics
-            .params
-            .iter()
-            .rev()
-            .take_while(|param| match param.kind {
-                ty::GenericParamDefKind::Lifetime => false,
-                ty::GenericParamDefKind::Type { has_default, .. } => {
-                    has_default
-                        && substs[param.index as usize]
-                            == GenericArg::from(
-                                self.tcx().bound_type_of(param.def_id).subst(self.tcx(), substs),
-                            )
-                }
-                ty::GenericParamDefKind::Const { has_default } => {
-                    has_default
-                        && substs[param.index as usize]
-                            == GenericArg::from(self.tcx().const_param_default(param.def_id))
-                }
-            })
-            .count();
-
-        &substs[own_params]
     }
 
     fn default_print_impl_path(

--- a/compiler/rustc_middle/src/ty/print/pretty.rs
+++ b/compiler/rustc_middle/src/ty/print/pretty.rs
@@ -825,12 +825,11 @@ pub trait PrettyPrinter<'tcx>:
 
         for (fn_once_trait_ref, entry) in fn_traits {
             // Get the (single) generic ty (the args) of this FnOnce trait ref.
-            let generics = self.generic_args_to_print(
-                self.tcx().generics_of(fn_once_trait_ref.def_id()),
-                fn_once_trait_ref.skip_binder().substs,
-            );
+            let generics = self.tcx().generics_of(fn_once_trait_ref.def_id());
+            let args =
+                generics.own_substs_no_defaults(self.tcx(), fn_once_trait_ref.skip_binder().substs);
 
-            match (entry.return_ty, generics[0].expect_ty()) {
+            match (entry.return_ty, args[0].expect_ty()) {
                 // We can only print `impl Fn() -> ()` if we have a tuple of args and we recorded
                 // a return type.
                 (Some(return_ty), arg_tys) if matches!(arg_tys.kind(), ty::Tuple(_)) => {
@@ -892,15 +891,13 @@ pub trait PrettyPrinter<'tcx>:
                 print(trait_ref.skip_binder().print_only_trait_name())
             );
 
-            let generics = self.generic_args_to_print(
-                self.tcx().generics_of(trait_ref.def_id()),
-                trait_ref.skip_binder().substs,
-            );
+            let generics = self.tcx().generics_of(trait_ref.def_id());
+            let args = generics.own_substs_no_defaults(self.tcx(), trait_ref.skip_binder().substs);
 
-            if !generics.is_empty() || !assoc_items.is_empty() {
+            if !args.is_empty() || !assoc_items.is_empty() {
                 let mut first = true;
 
-                for ty in generics {
+                for ty in args {
                     if first {
                         p!("<");
                         first = false;
@@ -1071,10 +1068,10 @@ pub trait PrettyPrinter<'tcx>:
                     let dummy_cx = cx.tcx().mk_ty_infer(ty::FreshTy(0));
                     let principal = principal.with_self_ty(cx.tcx(), dummy_cx);
 
-                    let args = cx.generic_args_to_print(
-                        cx.tcx().generics_of(principal.def_id),
-                        principal.substs,
-                    );
+                    let args = cx
+                        .tcx()
+                        .generics_of(principal.def_id)
+                        .own_substs_no_defaults(cx.tcx(), principal.substs);
 
                     // Don't print `'_` if there's no unerased regions.
                     let print_regions = args.iter().any(|arg| match arg.unpack() {

--- a/src/test/ui/const-generics/defaults/rp_impl_trait_fail.rs
+++ b/src/test/ui/const-generics/defaults/rp_impl_trait_fail.rs
@@ -15,12 +15,12 @@ impl Traitor<1, 2> for u64 {}
 
 
 fn uwu<const N: u8>() -> impl Traitor<N> {
-    //~^ error: the trait bound `u32: Traitor<N, N>` is not satisfied
+    //~^ error: the trait bound `u32: Traitor<N>` is not satisfied
     1_u32
 }
 
 fn owo() -> impl Traitor {
-    //~^ error: the trait bound `u64: Traitor<1_u8, 1_u8>` is not satisfied
+    //~^ error: the trait bound `u64: Traitor` is not satisfied
     1_u64
 }
 

--- a/src/test/ui/const-generics/defaults/rp_impl_trait_fail.stderr
+++ b/src/test/ui/const-generics/defaults/rp_impl_trait_fail.stderr
@@ -6,21 +6,21 @@ LL | fn rawr() -> impl Trait {
    |
    = help: the trait `Trait` is implemented for `Uwu<N>`
 
-error[E0277]: the trait bound `u32: Traitor<N, N>` is not satisfied
+error[E0277]: the trait bound `u32: Traitor<N>` is not satisfied
   --> $DIR/rp_impl_trait_fail.rs:17:26
    |
 LL | fn uwu<const N: u8>() -> impl Traitor<N> {
-   |                          ^^^^^^^^^^^^^^^ the trait `Traitor<N, N>` is not implemented for `u32`
+   |                          ^^^^^^^^^^^^^^^ the trait `Traitor<N>` is not implemented for `u32`
    |
    = help: the following other types implement trait `Traitor<N, M>`:
              <u32 as Traitor<N, 2_u8>>
              <u64 as Traitor<1_u8, 2_u8>>
 
-error[E0277]: the trait bound `u64: Traitor<1_u8, 1_u8>` is not satisfied
+error[E0277]: the trait bound `u64: Traitor` is not satisfied
   --> $DIR/rp_impl_trait_fail.rs:22:13
    |
 LL | fn owo() -> impl Traitor {
-   |             ^^^^^^^^^^^^ the trait `Traitor<1_u8, 1_u8>` is not implemented for `u64`
+   |             ^^^^^^^^^^^^ the trait `Traitor` is not implemented for `u64`
    |
    = help: the following other types implement trait `Traitor<N, M>`:
              <u32 as Traitor<N, 2_u8>>

--- a/src/test/ui/const-generics/defaults/trait_objects_fail.rs
+++ b/src/test/ui/const-generics/defaults/trait_objects_fail.rs
@@ -26,5 +26,5 @@ fn main() {
     foo(&10_u32);
     //~^ error: the trait bound `u32: Trait` is not satisfied
     bar(&true);
-    //~^ error: the trait bound `bool: Traitor<{_: u8}, {_: u8}>` is not satisfied
+    //~^ error: the trait bound `bool: Traitor<{_: u8}>` is not satisfied
 }

--- a/src/test/ui/const-generics/defaults/trait_objects_fail.stderr
+++ b/src/test/ui/const-generics/defaults/trait_objects_fail.stderr
@@ -9,16 +9,16 @@ LL |     foo(&10_u32);
    = help: the trait `Trait<2_u8>` is implemented for `u32`
    = note: required for the cast to the object type `dyn Trait`
 
-error[E0277]: the trait bound `bool: Traitor<{_: u8}, {_: u8}>` is not satisfied
+error[E0277]: the trait bound `bool: Traitor<{_: u8}>` is not satisfied
   --> $DIR/trait_objects_fail.rs:28:9
    |
 LL |     bar(&true);
-   |     --- ^^^^^ the trait `Traitor<{_: u8}, {_: u8}>` is not implemented for `bool`
+   |     --- ^^^^^ the trait `Traitor<{_: u8}>` is not implemented for `bool`
    |     |
    |     required by a bound introduced by this call
    |
    = help: the trait `Traitor<2_u8, 3_u8>` is implemented for `bool`
-   = note: required for the cast to the object type `dyn Traitor<{_: u8}, {_: u8}>`
+   = note: required for the cast to the object type `dyn Traitor<{_: u8}>`
 
 error: aborting due to 2 previous errors
 

--- a/src/test/ui/inference/need_type_info/concrete-impl.rs
+++ b/src/test/ui/inference/need_type_info/concrete-impl.rs
@@ -1,0 +1,16 @@
+trait Ambiguous<A> {
+    fn method() {}
+}
+
+struct One;
+struct Two;
+struct Struct;
+
+impl Ambiguous<One> for Struct {}
+impl Ambiguous<Two> for Struct {}
+
+fn main() {
+    <Struct as Ambiguous<_>>::method();
+    //~^ ERROR type annotations needed
+    //~| ERROR type annotations needed
+}

--- a/src/test/ui/inference/need_type_info/concrete-impl.stderr
+++ b/src/test/ui/inference/need_type_info/concrete-impl.stderr
@@ -1,0 +1,33 @@
+error[E0282]: type annotations needed
+  --> $DIR/concrete-impl.rs:13:5
+   |
+LL |     <Struct as Ambiguous<_>>::method();
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type of the type parameter `Self` declared on the trait `Ambiguous`
+   |
+help: consider specifying the generic argument
+   |
+LL |     <Struct as Ambiguous::<_>>::method();
+   |                         ~~~~~
+
+error[E0283]: type annotations needed
+  --> $DIR/concrete-impl.rs:13:5
+   |
+LL |     <Struct as Ambiguous<_>>::method();
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type of the type parameter `Self` declared on the trait `Ambiguous`
+   |
+note: multiple `impl`s satisfying `Struct: Ambiguous<_>` found
+  --> $DIR/concrete-impl.rs:9:1
+   |
+LL | impl Ambiguous<One> for Struct {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+LL | impl Ambiguous<Two> for Struct {}
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+help: consider specifying the generic argument
+   |
+LL |     <Struct as Ambiguous::<_>>::method();
+   |                         ~~~~~
+
+error: aborting due to 2 previous errors
+
+Some errors have detailed explanations: E0282, E0283.
+For more information about an error, try `rustc --explain E0282`.

--- a/src/test/ui/inference/need_type_info/self-ty-in-path.rs
+++ b/src/test/ui/inference/need_type_info/self-ty-in-path.rs
@@ -1,0 +1,13 @@
+// Test that we don't ICE when encountering a `Self` in a path.
+struct TestErr<T>(T);
+
+impl<T> TestErr<T> {
+    fn func_a<U>() {}
+
+    fn func_b() {
+        Self::func_a();
+        //~^ ERROR type annotations needed
+    }
+}
+
+fn main() {}

--- a/src/test/ui/inference/need_type_info/self-ty-in-path.stderr
+++ b/src/test/ui/inference/need_type_info/self-ty-in-path.stderr
@@ -1,0 +1,14 @@
+error[E0282]: type annotations needed
+  --> $DIR/self-ty-in-path.rs:8:9
+   |
+LL |         Self::func_a();
+   |         ^^^^^^^^^^^^ cannot infer type of the type parameter `U` declared on the associated function `func_a`
+   |
+help: consider specifying the generic argument
+   |
+LL |         Self::func_a::<U>();
+   |                     +++++
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0282`.

--- a/src/test/ui/inference/need_type_info/type-alias-indirect.rs
+++ b/src/test/ui/inference/need_type_info/type-alias-indirect.rs
@@ -1,0 +1,18 @@
+// An addition to the `type-alias.rs` test,
+// see the FIXME in that file for why this test
+// exists.
+//
+// If there is none, feel free to remove this test
+// again.
+struct Ty<T>(T);
+impl<T> Ty<T> {
+    fn new() {}
+}
+
+type IndirectAlias<T> = Ty<Box<T>>;
+fn indirect_alias() {
+    IndirectAlias::new();
+    //~^ ERROR type annotations needed
+}
+
+fn main() {}

--- a/src/test/ui/inference/need_type_info/type-alias-indirect.stderr
+++ b/src/test/ui/inference/need_type_info/type-alias-indirect.stderr
@@ -1,0 +1,9 @@
+error[E0282]: type annotations needed
+  --> $DIR/type-alias-indirect.rs:14:5
+   |
+LL |     IndirectAlias::new();
+   |     ^^^^^^^^^^^^^ cannot infer type for type parameter `T` declared on the type alias `IndirectAlias`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0282`.

--- a/src/test/ui/inference/need_type_info/type-alias.rs
+++ b/src/test/ui/inference/need_type_info/type-alias.rs
@@ -1,0 +1,36 @@
+// Test the inference errors in case the relevant path
+// uses a type alias.
+//
+// Regression test for #97698.
+struct Ty<T>(T);
+impl<T> Ty<T> {
+    fn new() {}
+}
+
+type DirectAlias<T> = Ty<T>;
+fn direct_alias() {
+    DirectAlias::new()
+    //~^ ERROR type annotations needed
+}
+
+type IndirectAlias<T> = Ty<Box<T>>;
+fn indirect_alias() {
+    IndirectAlias::new();
+    // FIXME: This should also emit an error.
+    //
+    // Added it separately as `type-alias-indirect.rs`
+    // where it does error.
+}
+
+struct TyDefault<T, U = u32>(T, U);
+impl<T> TyDefault<T> {
+    fn new() {}
+}
+
+type DirectButWithDefaultAlias<T> = TyDefault<T>;
+fn direct_but_with_default_alias() {
+    DirectButWithDefaultAlias::new();
+    //~^ ERROR type annotations needed
+}
+
+fn main() {}

--- a/src/test/ui/inference/need_type_info/type-alias.stderr
+++ b/src/test/ui/inference/need_type_info/type-alias.stderr
@@ -1,0 +1,15 @@
+error[E0282]: type annotations needed
+  --> $DIR/type-alias.rs:12:5
+   |
+LL |     DirectAlias::new()
+   |     ^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T`
+
+error[E0282]: type annotations needed
+  --> $DIR/type-alias.rs:32:5
+   |
+LL |     DirectButWithDefaultAlias::new();
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot infer type for type parameter `T`
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0282`.

